pallets/pallet-auction/src/lib.rs

```diff
@@ -0,0 +1,355 @@
+
+					if new_expiry > auction_data.expiry_block {
+						// The old auction id will be handled at on-finalize.
+						auction_data.expiry_block = new_expiry;
+
+						// Add the new block number to the storage.
+						AuctionsExpiryBlock::<T>::append(new_expiry, auction_id);
+					}
+
+					Self::deposit_event(Event::<T>::BidRegistered {
+						new_bidder,
+						auction_id,
+						new_price,
+					});
+
+					Ok(())
+				} else {
+					Err(Error::<T>::InvalidAuctionId.into())
+				}
+			})
+		}
+
+		#[pallet::call_index(2)]
+		#[pallet::weight(T::WeightInfo::cancel_auction())]
+		pub fn cancel_auction(origin: OriginFor<T>, auction_id: AuctionId) -> DispatchResult {
+			// Get the account id
+			let id = ensure_signed(origin)?;
+
+			Self::do_cancel_auction(auction_id, Some(&id))
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// Get the auction id to store into the NextAuctionId.
+		pub(crate) fn next_auction_id() -> AuctionId {
+			NextAuctionId::<T>::mutate(|id| {
+				*id = id.wrapping_add(1);
+				*id
+			})
+		}
+
+		/// Get the auction done.
+		fn resolve_auction(auction_id: AuctionId, auction_data: AuctionData<T>) {
+			if let Some((bider, price)) = auction_data.current_bid {
+				let bid_pool = T::BidsPoolAccount::get();
+
+				if price >= auction_data.reserve.unwrap_or_default() {
+					// transfer money
+					let tax = T::AuctionSuccessFeePercentage::get() * price;
+					let rest = price.saturating_sub(tax);
+					if let Ok(treasury) = T::Bank::treasury() {
+						let _ = T::Bank::transfer(&bid_pool, &treasury, tax);
+					}
+					// nft change owner, every Nft must have an owner, so that it must be ok.
+					if let Ok(owner) = T::NftManager::nft_transfer(auction_data.nft_id, &bider) {
+						let _ = T::Bank::transfer(&bid_pool, &owner, rest);
+					}
+					Self::deposit_event(Event::<T>::AuctionSucceeded {
+						auction_id,
+						to: bider,
+						asset: auction_data.nft_id,
+						price,
+					});
+				} else {
+					// return money to last bider
+					let _ = T::Bank::transfer(&bid_pool, &bider, price);
+					Self::deposit_event(Event::<T>::AuctionExpired { auction_id });
+				}
+			} else {
+				Self::deposit_event(Event::<T>::AuctionExpired { auction_id });
+			}
+
+			// nft change state
+			let _ = T::NftManager::change_nft_state(auction_data.nft_id, NftState::Free);
+		}
+
+		fn do_cancel_auction(auction_id: AuctionId, id: Option<&T::AccountId>) -> DispatchResult {
+			// Read the storage auctions, remove the auction id and return money to bidder.
+
+			if let Some(auction_data) = Auctions::<T>::take(auction_id) {
+				// If force cancel the id is none, only normal cancel will check owner of nft.
+				if let Some(owner) = id {
+					T::NftManager::ensure_nft_owner(owner, auction_data.nft_id)?;
+				}
+
+				// Only normal cancel will compare whether current price is upon the reserve price.
+				if let Some((bider, price)) = auction_data.current_bid {
+					if id.is_some() && price >= auction_data.reserve.unwrap_or_default() {
```

Comment:
As a learning excercise, try to use
(bool && bool) -> Option -> Result ?;
format

Reply:
Same goes the above
if let Some(auction_data) = Auctions::<T>::take(auction_id)
as well